src/vision_overhaul.cpp

```diff
@@ -0,0 +1,455 @@
+    if (temp_room->light[2]) {
+      if (tn_with_thermo > 0) {
+        int new_thermo = MAX(0, tn_with_thermo - temp_room->light[2]);
+        if (new_thermo != tn_with_thermo) {
+          TH_PENALTY("LightSpell[TH]", new_thermo - tn_with_thermo);
+        }
+      }
+
+      if (tn_with_ll > 0) {
+        int new_ll = MAX(0, tn_with_ll - temp_room->light[2]);
+        if (new_ll != tn_with_ll) {
+          LL_PENALTY("LightSpell[LL]", new_ll - tn_with_ll);
+        }
+      }
+
+      if (tn_with_none > 0) {
+        int new_normal = MAX(0, tn_with_none - temp_room->light[2]);
+        if (new_normal != tn_with_none) {
+          NM_PENALTY("LightSpell[NM]", new_normal - tn_with_none);
+        }
+      }
+    } else {
+      // This happens when you roll no successes at all, right?
+      // light_target /= 2;
+      mudlog("Encountered that weird light case, investigate here!", ch, LOG_SYSLOG, TRUE);
+    }
+  }
+
+  if (temp_room->vision[1] == LIGHT_MIST) {
+    if (!has_natural_lowlight_vision) {
+      LL_PENALTY("Mist[LL-U]", 2);
+    }
+
+    NM_PENALTY("Mist[NM]", 2);
+  }
+
+  if (temp_room->vision[1] == LIGHT_LIGHTSMOKE
+      || (weather_info.sky == SKY_RAINING && temp_room->sector_type != SPIRIT_HEARTH && !ROOM_FLAGGED(temp_room, ROOM_INDOORS)))
+  {
+    if (has_natural_lowlight_vision) {
+      LL_PENALTY("LSmoke/LRain[LL-N]", 2);
+    } else {
+      LL_PENALTY("LSmoke/LRain[LL-U]", 4);
+    }
+
+    NM_PENALTY("LSmoke/LRain[NM]", 4);
+  }
+
+  if (temp_room->vision[1] == LIGHT_HEAVYSMOKE
+      || (weather_info.sky == SKY_LIGHTNING && temp_room->sector_type != SPIRIT_HEARTH && !ROOM_FLAGGED(temp_room, ROOM_INDOORS)))
+  {
+    if (!has_natural_thermographic_vision) {
+      TH_PENALTY("HSmoke/HRain[TH-U]", 1);
+    }
+
+    if (has_natural_lowlight_vision) {
+      LL_PENALTY("HSmoke/HRain[LL-N]", 4);
+    } else {
+      LL_PENALTY("HSmoke/HRain[LL-U]", 6);
+    }
+
+    NM_PENALTY("HSmoke/HRain[NM]", 6);
+  }
+
+  if (temp_room->vision[1] == LIGHT_THERMALSMOKE) {
+    if (has_natural_thermographic_vision) {
+      TH_PENALTY("TSmoke[TH-N]", 6);
+    } else {
+      TH_PENALTY("TSmoke[TH-U]", 8);
+    }
+
+    LL_PENALTY("TSmoke[LL]", 4);
+
+    NM_PENALTY("TSmoke[NM]", 4);
+  }
+
+  // We've reached the end of all the penalties. We'll hand back the lowest.
+  std::map<const char *, int> *penalty_vector;
+  int penalty_chosen;
+
+  if (has_vision(ch, VISION_THERMOGRAPHIC) && tn_with_thermo <= tn_with_ll && tn_with_thermo <= tn_with_none) {
+    penalty_vector = &thermo_penalties;
+    penalty_chosen = tn_with_thermo;
+  }
+  else if (has_vision(ch, VISION_LOWLIGHT) && tn_with_ll <= tn_with_thermo && tn_with_ll <= tn_with_none) {
+    penalty_vector = &lowlight_penalties;
+    penalty_chosen = tn_with_ll;
+  }
+  else {
+    penalty_vector = &normal_penalties;
+    penalty_chosen = tn_with_none;
+  }
+
+  for (auto it = penalty_vector->begin(); it != penalty_vector->end(); ++it) {
```

Comment:
Don't we have access to range-based for loops here? `for (auto it : penalty_vector)`